timelock-server/src/integTest/java/com/palantir/atlasdb/timelock/MultiNodePaxosTimeLockServerIntegrationTest.java

```diff
@@ -0,0 +1,143 @@
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+
+import com.google.common.collect.ImmutableSortedMap;
+import com.palantir.lock.LockClient;
+import com.palantir.lock.LockMode;
+import com.palantir.lock.LockRefreshToken;
+import com.palantir.lock.LockRequest;
+import com.palantir.lock.StringLockDescriptor;
+
+import feign.RetryableException;
+
+public class MultiNodePaxosTimeLockServerIntegrationTest {
+    private static final String CLIENT_1 = "test";
+
+    private static final TestableTimelockCluster CLUSTER = new TestableTimelockCluster(
+            "http://localhost",
+            CLIENT_1,
+            "paxosMultiServer0.yml",
+            "paxosMultiServer1.yml",
+            "paxosMultiServer2.yml");
+
+    private static final String LOCK_CLIENT = LockClient.ANONYMOUS.getClientId();
+    private static final LockRequest LOCK_REQUEST = LockRequest.builder(
+            ImmutableSortedMap.of(
+                    StringLockDescriptor.of("lock1"),
+                    LockMode.WRITE))
+            .doNotBlock()
+            .build();
+
+    @ClassRule
+    public static final RuleChain ruleChain = CLUSTER.getRuleChain();
+
+    @BeforeClass
+    public static void waitForClusterToStabilize() {
+        CLUSTER.waitUntilLeaderIsElected();
+    }
+
+    @Before
+    public void bringAllNodesOnline() {
+        CLUSTER.waitUntillAllSeversAreOnlineAndLeaderIsElected();
+    }
+
+    @Test
+    public void non_leaders_return_503() {
+        CLUSTER.nonLeaders().forEach(server -> {
+            assertThatThrownBy(() -> server.getFreshTimestamp())
+                    .isInstanceOf(RetryableException.class);
+            assertThatThrownBy(() -> server.lock(LOCK_CLIENT, LOCK_REQUEST))
+                    .isInstanceOf(RetryableException.class);
+        });
+    }
+
+    @Test
+    public void leader_responds_to_requests() throws InterruptedException {
+        CLUSTER.currentLeader().getFreshTimestamp();
+
+        LockRefreshToken token = CLUSTER.currentLeader().lock(LOCK_CLIENT, LOCK_REQUEST);
+        CLUSTER.unlock(token);
+    }
+
+    @Test
+    public void new_leader_takes_over_if_current_leader_dies() {
+        CLUSTER.currentLeader().kill();
+
+        CLUSTER.getFreshTimestamp();
+    }
+
+    @Test
+    public void leader_loses_leadership_if_a_quorum_is_not_alive() {
+        TestableTimelockServer leader = CLUSTER.currentLeader();
+        CLUSTER.nonLeaders().forEach(TestableTimelockServer::kill);
+
+        assertThatThrownBy(() -> leader.getFreshTimestamp())
+                .isInstanceOf(RetryableException.class);
+    }
+
+    @Test
+    public void someone_becomes_leader_again_after_quorum_is_restored() {
+        CLUSTER.nonLeaders().forEach(TestableTimelockServer::kill);
+        CLUSTER.nonLeaders().forEach(TestableTimelockServer::start);
+
+        CLUSTER.getFreshTimestamp();
+    }
+
+    @Test
+    public void can_perform_rolling_restart() {
```

Comment:
Discussed offline with @jeremyk-91 - I was unable to repro this locally after many runs. I think we can go ahead and merge, and will address if we see flakes.